### ISSUE TITLE
"corner_peaks" Function not Handling Image Borders Correctly

### ISSUE THREAD

**Issue by Tea7ime** [scikit-image/scikit-image#4200]
I am using the corner peaks function to count the number of objects in an image. I was noticing that it is returning too many peaks close to image borders. As the min_distance parameter is increased, the number of peaks found also increases (along borders). 

I don't believe this is the expected result? I would expect corner_peaks to find only one peak for each of the edge groups with 'min_distance=2'. 

![peak_corners_incorrect](https://user-images.githubusercontent.com/16522814/65978200-44cfff80-e441-11e9-8cb6-49a420041a9d.png)

Expected Result:

![peak_corners_expected](https://user-images.githubusercontent.com/16522814/65978894-77c6c300-e442-11e9-8c90-b8935d3c9d2e.png)

While troubleshooting, I'm seeing that the "skimage.feature.corner_peaks" is trying to set negative index locations on the image to false. When it tries to do this, nothing is actually changed. I believe it should be checking for negative indexes and setting them to 0. 

 For the example below, here are the problem index locations corner_peaks is trying to set to false.

Setting this location to False: [-2: 3,  -2: 3]
Setting this location to False: [-2: 3,  7: 12]
setting this location to False: [2: 7,  -2: 3]
setting this location to False: [6: 11,  -2: 3]

This is why all the peaks along the left side and top are not being reduced to a single point.

I was hoping to contribute a fix for this if you agree that this is an issue? I was able to solve the problem by checking for negatives. 

## Way to reproduce

To recreate unexpected result:

```python
import numpy as np
import matplotlib.pyplot as plt
from skimage.feature import corner_peaks

img = np.zeros((10, 10), dtype=np.float32)
img[0:2, 0] = 1.0
img[1, 1] = 0.5

img[0:2, 9] = 0.8
img[1, 8] = 0.7

img[2:4, 3] = 1.0
img[3, 4] = 0.5

img[2:4, 6] = 1.0
img[3, 5] = 0.5

img[4:6, 9] = 0.7
img[4:6, 0] = 0.7

img[6:8, 3] = 1.0
img[6, 4] = 0.5

img[6:8, 6] = 1.0
img[6, 5] = 0.5

img[8:10, 9] = .9
img[8,8] = 0.5

img[8:10, 0] = 0.6
img[8,1] = 0.5

peaks = corner_peaks(img, min_distance=2, exclude_border=False)

plt.imshow(img, cmap='gray')
plt.plot(peaks[:, 1], peaks[:, 0], 'bo', markersize=2)
plt.show()
```

## Version information
```python
3.7.0 (default, Jun 28 2018, 07:39:16) 
[Clang 4.0.1 (tags/RELEASE_401/final)]
Darwin-18.7.0-x86_64-i386-64bit
scikit-image version: 0.15.0
numpy version: 1.16.4
```

Thank you!

**Comment by jni** [comment]
@Tea7ime fascinating! Thanks for contributing and I agree with your expected result and your diagnosis. A fix together with your test case would be most welcome!

Honestly though, I'm not sure why this function is called `corner_peaks`...

**Comment by Tea7ime** [comment]
@jni Thanks for reviewing this! I'll get a pull request together soon. :-)

Fair question lol. I suppose it has something to do with how it ends up leaving the last peak in an area? The last peak is always the top left corner of a true area in the mask(retrieved from peak_local_max).

**Comment by lagru** [comment]
Solved in #4207.
